### PR TITLE
Context menu to toggle error bars on plot

### LINKED ISSUE #25547: Workbench: Quick option to enable/disable error bars on a plot
There is currently no way to enable/disable error bars on an existing 1D plot. This issue is to allow a user to enable/disable error bars quickly for all or some curves on a plot.

For the all curves option we will need to think whether this is best as a context-menu option or a more visible button. For individual curves it may make sense to include this is the general config dialog options.

### PR DESCRIPTION
**Pull request by dtasev**
Will be wrapped up after ~~#25945~~ is merged

**Description of work.**
Adds context menu to toggle error bars on plot.

Errors for each line can be toggled individually, or all errors can be toggled/shown/hidden.

**To test:**
#### Setup
Script for **8** script plots and 2 workspaces.
```python
# The following line helps with future compatibility with Python 3
# print must now be used as a function, e.g print('Hello','World')
from __future__ import (absolute_import, division, print_function, unicode_literals)

# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *

import matplotlib.pyplot as plt

import numpy as np
ws2d_histo = CreateWorkspace(DataX=[10, 20, 30, 10, 20, 30, 10, 20, 30],
                             DataY=[2, 3, 4, 5, 3, 5],
                             DataE=[1, 2, 3, 4, 1, 1],
                             NSpec=3,
                             Distribution=True,
                             UnitX='Wavelength',
                             VerticalAxisUnit='DeltaE',
                             VerticalAxisValues=[4, 6, 8])

ws2d_1spec = CreateWorkspace(DataX=[10, 20, 30],
                                    DataY=[2, 3],
                                    DataE=[1, 2],
                                    NSpec=1,
                                    Distribution=True,
                                    UnitX='Wavelength',
                                    VerticalAxisUnit='DeltaE',
                                    VerticalAxisValues=[4])
                                    
def do_plots(bonus_kwargs=None):
    if bonus_kwargs is None:
        bonus_kwargs = {}
    plt.figure()
    plt.title('Script Plot without Errors')
    plt.plot([0,15000], [0,15000], label='MyLabel')
    plt.plot([0,15000], [0,14000], label='MyLabel 2')
    plt.legend()

    x = np.linspace(0, 15000, 150)
    plt.figure()
    plt.title('Script Plot with YErr and XErr.{}'.format('' if len(bonus_kwargs) == 0 else " Bonus ARGS: {}".format(bonus_kwargs)))
    plt.errorbar(x, 2*x, yerr=5*np.sqrt(2*x), xerr=2*np.sqrt(2*x), label='MyLabel', **bonus_kwargs)
    plt.legend()

    x = np.linspace(0, 15000, 150)
    plt.figure()
    plt.title('Script Plot with ONLY YErr.{}'.format('' if len(bonus_kwargs) == 0 else " Bonus ARGS: {}".format(bonus_kwargs)))
    plt.errorbar(x, 2*x, yerr=5*np.sqrt(2*x), label='MyLabel', **bonus_kwargs)
    plt.legend()

    x = np.linspace(0, 15000, 150)
    plt.figure()
    plt.title('Script Plot with ONLY XErr.{}'.format('' if len(bonus_kwargs) == 0 else " Bonus ARGS: {}".format(bonus_kwargs)))
    plt.errorbar(x, 2*x, xerr=2*np.sqrt(2*x), label='MyLabel', **bonus_kwargs)
    plt.legend()

do_plots()
do_plots({'capsize':4})
plt.show()
```

#### To Test
- `ws2d_1spec` - check that the error bar for the 1 spectra can be toggled from:
  - Right Click > Error Bars > Show/Hide all errors
  - Right Click > Error Bars > click on the 1 spectra shown
  - Cogwheel for plot options > Curves > Errorbars > check/uncheck `hide errorbars`
- `ws2d_histo`
  - Same as above, but for more than one curve
- Scripted plots
  - Check that Y and X errors are toggled correctly
    - Toggling the errors should change the state of both Y and X errors at the same time
  - Check that the one without errors doesn't provide the context menu at all

- Check that plots can be saved and loaded correctly, before and after toggling the errorbars.

Fixes #25547

Release notes added in `docs/source/release/v4.1.0/ui.rst`.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by ghost — CHANGES_REQUESTED** [review]
I get hard crashes when I try to toggle error bars on scripted plots. Maybe it's worth checking the axes is a MantidAxes before adding the options to the context menu?

To recreate:
```python
plt.figure()
plt.title('My Title')
plt.plot([0,15000], [0,15000], label='MyLabel')
plt.plot([0,15000], [0,14000], label='MyLabel 2')
plt.legend()
plt.show()
```
Right-click then select any errorbar options.

**Comment by DTasev** [comment]
Good spot! I've added the check. What should be done about scripted plots that  contain errors though? Should the menu still be shown?

**Comment by DTasev** [comment]
Another thing I've left in by accident is the show/hide all buttons in the context menu. My main reason for adding those was to test the functions as they will likely be used for the plot settings dialog's `Apply to All`. 

Should I remove those from the context menu?

**Comment by DTasev** [comment]
The scripted plots issue should be fixed. 

The context menu will now be shown _only_ if there is a line with errors plotted, and only work on that line's errors.

I haven't removed the show/hide all buttons yet.

**Comment by hsautessella** [comment]
Another issue is x-errors. If you plot something with x and y errors and toggle the errorbars only one set of the errorbars will toggle. We probably want a menu option to toggle x-errors as well if they're available.

A script for x and y errorbars to make your life easier:
```python 
x = np.linspace(0, 15000, 150)
plt.figure()
plt.title('My Title')
plt.errorbar(x, 2*x, yerr=5*np.sqrt(2*x), xerr=2*np.sqrt(2*x), label='MyLabel')
plt.legend()
plt.show()
```
Also if you now drag a workspace on top of that figure, the first figure is not included in the legend. It's not immediately clear why, but I've had a little dig around and it could be when we're converting the Axes to a MantidAxes i.e. `MantidAxes.from_mpl_axes` the line being passed somehow loses its label and defaults to `_no_legend_`.

**Comment by DTasev** [comment]
Should x-errors be a separate entry in the context menu, making it show 
```
Errors for Y -> all lines with errors on Y
Errors for X -> all lines with errors on X
```
Can lines have just the X error without Y errors?

**Comment by hsautessella** [comment]
I think a separate entry would be good, but definitely get a second opinion. I don't see why a curve couldn't have only its x-errors showing, though admittedly it will probably be a rare-use case.

**Comment by DTasev** [comment]
After an offline discussion with @martyngigg it was decided that the all the Y error related options should be in a submenu, now called `Y Error Bars`. 

Additionally, X errors are not shown/hidden when an action is performed. This could be a future extension, if required

**Comment by DTasev** [comment]
Is there another way I am supposed to clear plots between tests? [This way](https://github.com/mantidproject/mantid/pull/25617/files#diff-9635e88c1c41bcb30b4d61224b16022eR50) seems to leave lines hanging around which causes all the test failures

File: Framework/PythonInterface/mantid/plots/__init__.py
Comment (martyngigg) [inline Framework/PythonInterface/mantid/plots/__init__.py]: Could you help me understand the scenario where MPL appends the errors on the bottom of the legend? I haven't seen that before.

**Review by martyngigg — CHANGES_REQUESTED** [review]
Code changes look good. A nice set of tests.

First issue I found when testing was when I loaded a workspace, plotted a spectrum *without* errors and then clicked "Hide all errors" and I got the image below:

Screenshot from 2019-05-13 10-18-22

Another one was if I do a plot without errors then select show errors I get a set of error bars where the caps are different colours to the error bars:

Screenshot from 2019-05-13 10-23-08

File: Framework/PythonInterface/mantid/plots/__init__.py
Comment (DTasev) [reply Framework/PythonInterface/mantid/plots/__init__.py]: It seems that is the possibly unintended, but default behaviour when you have a mixture of lines that do not have `ErrorbarContainer`s associated with them, and lines that do. It happens because the `ax.containers` are added last [here](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/legend.py#L1212), which makes lines with errors always appear below lines without errors.

**Comment by DTasev** [comment]
Putting this to `In Progress` as I've found a couple of issues:
- Saving/loading project does not keep the caps lines' color. It also seems to make the references in the `ax.containers` wrong.
- `ax.creation_args` duplicates its contents on load -> I think when loaded back in and the correct `creation_args.function` is called, it adds another entry in the creation args.

**Comment by TTitcombe** [comment]
@DTasev I've fixed a couple of the bugs, however there are a few things that I think will need your input before this can continue, as I'm not familiar enough with the code: 
* there are a few commented out pieces of code you've introduced in previous commits. I'm not sure if these should be removed
* A few bugs were due to POST_CREATION_ARGS being assumed to exist. I've set default args at places where the errors were raised, however a better solution may be to set default POST_CREATION_ARGS with `None` values when creation args are created.
* There are a few merge conflicts I'm unsure how to deal with

Finally, I'm not convinced the issue of saving/loading mixing up the order of lines is an actual issue. Your instructions for replicating are:
* plot
* save project
* load project (while the plot is open)

When I closed the plot before re-opening, I saved/loaded several times without problem. I think the issue is when you have two plots open pointing to the same data. If that is the case, I'm not sure this would be a particularly important issue to fix.

File: qt/python/mantidqt/project/plotsloader.py
Comment (DTasev) [inline qt/python/mantidqt/project/plotsloader.py]: These 3 lines can be removed I think, the code below doesn't use it anymore

File: qt/python/mantidqt/project/plotsloader.py
Comment (DTasev) [inline qt/python/mantidqt/project/plotsloader.py]: I don't think I've removed the implementation of this function, it might not be used anymore, and this functionality is merged into the `for cargs in creation_args[0]:` above

File: qt/python/mantidqt/project/plotssaver.py
Comment (DTasev) [inline qt/python/mantidqt/project/plotssaver.py]: This was commented out as it was unused

File: Framework/PythonInterface/test/python/mantid/plots/plotfunctionsTest.py
Comment (DTasev) [inline Framework/PythonInterface/test/python/mantid/plots/plotfunctionsTest.py]: I think the lines above are the new values

File: qt/applications/workbench/workbench/plotting/figureerrorsmanager.py
Comment (DTasev) [inline qt/applications/workbench/workbench/plotting/figureerrorsmanager.py]: Is this missing a `.POST_CREATION_ARGS`?

**Review by dtasev — COMMENTED** [review]
>there are a few commented out pieces of code you've introduced in previous commits. I'm not sure if these should be removed

Hopefully I got all of them in the review

>A few bugs were due to POST_CREATION_ARGS being assumed to exist. I've set default args at places where the errors were raised, however a better solution may be to set default POST_CREATION_ARGS with None values when creation args are created.

This is sensible, but I think we still need to keep the check if `POST_CREATION_ARGS` exists otherwise plots saved in 4.0 might error after this PR.

I agree the saving/loading is lower priority than I previously thought

File: qt/applications/workbench/workbench/plotting/figureerrorsmanager.py
Comment (TTitcombe) [reply qt/applications/workbench/workbench/plotting/figureerrorsmanager.py]: Yes, good spot

File: qt/python/mantidqt/project/plotsloader.py
Comment (TTitcombe) [reply qt/python/mantidqt/project/plotsloader.py]: I've removed plot_extra_lines and the other comments

**Comment by TTitcombe** [comment]
@martyngigg @hsautessella @DTasev 
This is ready for another review

**Comment by martyngigg** [comment]
Thanks Tom!

File: qt/applications/workbench/workbench/plotting/figuremanager.py
Comment (DTasev) [inline qt/applications/workbench/workbench/plotting/figuremanager.py]: I think these may be `numpy.ndarray`. If so they should have `.min()`/`.max()` methods

File: qt/applications/workbench/workbench/plotting/figuremanager.py
Comment (DTasev) [inline qt/applications/workbench/workbench/plotting/figuremanager.py]: @TTitcombe I think MPL sets some margins on the limits, as there's always some space between the data and the borders of the plot. Currently, after showing the errors and going to the "home" view the data seems to end exactly at the border of the plot. 

Do you think it'd be possible to the margins back while errors are shown? I found [this](https://stackoverflow.com/a/42804403/2823526) which mentions this method `ax.set_ymargin(0.1)`, alternatively also gives this `ymax = data.max()+0.4*(data.max()-data.min())` for a 40% margin (I think about 10%-15% should be enough). Another possible route is to check how MPL does it, but that would involve some digging in the source

Comment (TTitcombe) [reply qt/applications/workbench/workbench/plotting/figuremanager.py]: Good idea. I've calculated the margin manually by adding padding by 5% of the range of the data to the top and bottom (`set_ymargin` didn't work)

**Comment by TTitcombe** [comment]
@DTasev I've included your comments, and I've fixed a case where a spectrum with errors plot was being scaled incorrectly when the errors had been hidden.

I've added a couple of tests to `test_figuremanager` to check that the correct plot bounds are being selected, however the test suite takes ~3s to run for only 4 tests. I think the issue is related to mocking out QAppThreadCall everytime. Is it necessary to include this in the tests?

If we want to expand the functionality of figuremanager, we'll have to figure out a way to test the functionality without including the thread-mocking overhead

File: Framework/PythonInterface/mantid/plots/utility.py
Comment (martyngigg) [reply Framework/PythonInterface/mantid/plots/utility.py]: Is this in fact used in the `mantid.plots` package? Should it belong in the `figureerrorsmanager`?

**Review by martyngigg — DISMISSED** [review]

**Comment by martyngigg** [comment]
@TTitcombe Do you think you would have time to look over the final few things here? If not I could get Harry to take a look.

**Comment by TTitcombe** [comment]
@martyngigg I can look at this

File: Framework/PythonInterface/mantid/plots/utility.py
Comment (TTitcombe) [reply Framework/PythonInterface/mantid/plots/utility.py]: It's used in `workbench.plotting`, `mantid.plots`, and `mantidqt.project`, so I'm not sure where the best place for it would be

**Comment by hsautessella** [comment]
I'm getting errors on this whenever I try to show errorbars.
To recreate:
- Run:
```python
ws2d_1spec = CreateWorkspace(DataX=[10, 20, 30],
                                    DataY=[2, 3],
                                    DataE=[1, 2],
                                    NSpec=1,
                                    Distribution=True,
                                    UnitX='Wavelength',
                                    VerticalAxisUnit='DeltaE',
                                    VerticalAxisValues=[4])
```
- `Right-click on workspace -> Plot -> Spectrum`. 
- `Right-click on plot -> Y Error Bars -> Show All Errors`. Gives me the error:

```
Traceback (most recent call last):
  File "c:\users\ejo73213\mantid\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 337, in _update_plot_after
    func(*args, **kwargs)
  File "c:\users\ejo73213\mantid\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 271, in toggle_all_error_bars
    self._add_errorbar_for(index, data_line, make_visible)
  File "c:\users\ejo73213\mantid\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 144, in _add_errorbar_for
    specNum = ax.creation_args[creation_args_index]["specNum"]
KeyError: 'specNum'
```
Also toggling normalization seems to break when after toggling errorbars

**Comment by martyngigg** [comment]
Are we lacking a developer on this now?

**Comment by martyngigg** [comment]
@DTasev Any chance you would have some time to polish this off?

**Comment by DTasev** [comment]
I'll have a look at this after #25945 is merged - I believe the functionality from it can be reused here

**Comment by martyngigg** [comment]
@DTasev #25945 has gone in now.

File: qt/python/mantidqt/project/test/test_plotssaver.py
Comment (DTasev) [inline qt/python/mantidqt/project/test/test_plotssaver.py]: This is just from formatting - I haven't changed any of the values

File: qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/presenter.py
Comment (hsautessella) [inline qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/presenter.py]: It looks like we're calling `legend` and calling `relim` twice here. Once in `replot_selected_curve` and then again in `toggle_errorbars`. We probably want to take `relim` and `legend` out of these functions and call them at the end of `apply_properties` and at the end of `FigureErrorsManager.toggle_errorbars_for`?

**Review by ghost — CHANGES_REQUESTED** [review]
This looks good and works like a charm. Just a couple of comments below 👍

File: qt/applications/workbench/workbench/plotting/figureerrorsmanager.py
Comment (hsautessella) [inline qt/applications/workbench/workbench/plotting/figureerrorsmanager.py]:
```suggestion
                # this might not have creation_args
```

File: qt/applications/workbench/workbench/plotting/figureerrorsmanager.py
Comment (hsautessella) [inline qt/applications/workbench/workbench/plotting/figureerrorsmanager.py]: I think because we're not converting the `hide` argument into the `visible` Matplotlib kwarg if you set a curve to hidden in the plot options, and then toggle errorbars, the curve will no longer be hidden. I don't know if this is desired behaviour or not, but if we do want to keep the curve hidden in this case, we just need this line to be changed to
```python
plot_kwargs = curve_props.get_plot_kwargs()
```

**Comment by martyngigg** [comment]
I'll leave this one to Harry.

File: qt/applications/workbench/workbench/plotting/figuremanager.py
Comment (DTasev) [inline qt/applications/workbench/workbench/plotting/figuremanager.py]: leftover code

File: qt/applications/workbench/workbench/plotting/test/test_figureerrorsmanager.py
Comment (DTasev) [reply qt/applications/workbench/workbench/plotting/test/test_figureerrorsmanager.py]: This is a bit sneaky but `MantidAxes` is used by the `type` specification [here](https://github.com/mantidproject/mantid/pull/25617/files/7a82ace5caf98a3a1ce5f3db7c7f1b5fe4e247ab#diff-9635e88c1c41bcb30b4d61224b16022eR129), as well as the projection registration. I think it's worth the comment nonetheless

File: qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/__init__.py
Comment (DTasev) [reply qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/__init__.py]: :+1: Good spot, it's not used anywhere anymore!

File: qt/applications/workbench/workbench/plotting/figureerrorsmanager.py
Comment (DTasev) [reply qt/applications/workbench/workbench/plotting/figureerrorsmanager.py]: I agree it doesn't feel right, and I have opened a maintenance task to refactor it: https://github.com/mantidproject/mantid/issues/26330

File: qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/presenter.py
Comment (DTasev) [reply qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/presenter.py]: I've refactored the code that decides how to replot the curve into another function, which is now reused from both the `CurvesTabWidgetPresenter` and `FigureErrorsManager`. `apply_properties` now uses that. Additionally the relim/scale/legend is now in `update_limits_and_legend` called from both classes separately

File: qt/applications/workbench/workbench/plotting/figureerrorsmanager.py
Comment (DTasev) [reply qt/applications/workbench/workbench/plotting/figureerrorsmanager.py]: I am not sure about the behaviour, but this lets me reuse the same code as the presenter, which I'm all for

File: qt/applications/workbench/workbench/plotting/test/test_figureerrorsmanager.py
Comment (DTasev) [reply qt/applications/workbench/workbench/plotting/test/test_figureerrorsmanager.py]: Also I didn't know I could just stick in a `# noqa` and the lines below will be unaffected - very nice!

**Comment by DTasev** [comment]
@hsautessella Ready for another review

**Comment by DTasev** [comment]
@martyngigg @NickDraper The issue #25547 for this has been re-targeted for 4.2, shall we still aim to get this in 4.1 in this case?

**Comment by martyngigg** [comment]
@DTasev Yes, please pull the issue back

File: qt/applications/workbench/workbench/plotting/test/test_figureerrorsmanager.py
Comment (DTasev) [reply qt/applications/workbench/workbench/plotting/test/test_figureerrorsmanager.py]: Nevermind! Flake8 didn't recognise the type usage :(

File: qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/presenter.py
Comment (hsautessella) [inline qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/presenter.py]: I think you may have forgotten to take this function out

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by hsautessella** [comment]
This looks good to me 👍

**Review by ghost — APPROVED** [review]

**Comment by DTasev** [comment]
There's some leftover debug code in `figuremanager.py`, I'm taking a look

**Review by NickDraper — CHANGES_REQUESTED** [review]
Release notes are in the wrong place and a resurrecting a dead file that had otherwise been removed.  Please move the line to the workbench.rst.

No need to wait for the builds after this, just give me a nudge and I'll merge it.

**Comment by NickDraper** [comment]
I made the release note changes

**Pull request merged**
